### PR TITLE
Fix hybrid bug, double performance

### PR DESCRIPTION
**Pull request by dellaert**
I could not really separate the functional pruning PR from this one. Here's everything in this PR:

## Fixing the bug
- BIG: fix bug in HybridConditional::errorTree and HGFG::errorTree
- made errorTree much faster by calling polymorphic errorTree and using AlgebraicDecisionTree operators

## `discretePosterior`
- renamed both `HBN::evaluate` and `HGFG::probPrime` that take VectorValues to `discretePosterior`: these are very powerful. 

## Simplified API 
- removed `HBN::logProbability`: errorTree is all we need and this one has confusing semantics
- removed `HGC::logProbability` that takes continuous values: very problematic name and only used in one place: again, errorTree is all we need.

## Functional pruning
- pruning is now functional
- got rid of duplicate (unused) prunerFunc
- added a write-access conditional to allow iSAM versions to be still imperative

## Update Oct 1:
- Reworked `HGC::prune` to use `ADT::max`
- Rewrote `HybridBayesNet::prune` completely

Looking at pruned trees for testHybridEstimation is very informative: it seems that the ordering of the keys is working against us in terms of pruning: we have large duplicate trees:

step 2:
image

step 5:
image

...

step 14:
image

### CODE REVIEW

**Comment by dellaert** [comment]
Note: “doubling performance” refers to testHybridEstimation. Could be twice as fast because now the discrete factors are properly taken into account when pruning.

File: gtsam/hybrid/HybridConditional.cpp
Comment (dellaert) [inline gtsam/hybrid/HybridConditional.cpp:133]: Bug 1

File: gtsam/hybrid/HybridGaussianFactorGraph.cpp
Comment (dellaert) [inline gtsam/hybrid/HybridGaussianFactorGraph.cpp:516]: Bug 2

File: gtsam/hybrid/HybridBayesNet.cpp
Comment (varunagrawal) [inline gtsam/hybrid/HybridBayesNet.cpp]: Why a copy if everything is functional?

File: gtsam/hybrid/HybridBayesNet.h
Comment (varunagrawal) [inline gtsam/hybrid/HybridBayesNet.h]: Should be `log p(x)` not `log P(x)`

File: gtsam/hybrid/tests/testHybridNonlinearFactorGraph.cpp
Comment (varunagrawal) [inline gtsam/hybrid/tests/testHybridNonlinearFactorGraph.cpp:1005]: I kind of liked these regression tests.

**Review by varunagrawal — APPROVED** [review]
Some questions, especially about a copy.

File: gtsam/hybrid/HybridBayesNet.cpp
Comment (dellaert) [reply gtsam/hybrid/HybridBayesNet.cpp]: You're right, it was left-over ! I'll fix

Comment (dellaert) [reply gtsam/hybrid/HybridBayesNet.cpp]: Actually, I messed up the discrete part. Did you ever profile the pruning as it is? It's an exponential operation - that alone might be responsible for slowness in some experiments.

I'll fix but we have to chat about this more I think.

File: gtsam/hybrid/tests/testHybridNonlinearFactorGraph.cpp
Comment (dellaert) [reply gtsam/hybrid/tests/testHybridNonlinearFactorGraph.cpp:1005]: I will replace them with better, non-regression test based on discreteConditional.

File: gtsam/hybrid/HybridConditional.cpp
Comment (varunagrawal) [reply gtsam/hybrid/HybridConditional.cpp:133]: Oh interesting, I was missing the `DiscreteFactor::errorTree` method. There's actually an improvement here I can make.

File: gtsam/hybrid/HybridBayesNet.cpp
Comment (varunagrawal) [reply gtsam/hybrid/HybridBayesNet.cpp]: I haven't. I would like to see your perf-profiling setup and make it standard across projects. :) Seems to be very useful.

Comment (dellaert) [reply gtsam/hybrid/HybridBayesNet.cpp]: I'm on a mac. I just build the binary and use Instruments CPU profiler.

Comment (dellaert) [reply gtsam/hybrid/HybridBayesNet.cpp]: OK, I fixed it. 

@varunagrawal :
- You probably had pruning tests in your project, but I added some here so I could refactor HGC::prune based on ADT::max. 
- You might want to check out his branch and try whether your experiments still work before we merge?

**Comment by dellaert** [comment]
@varunagrawal please merge this after you established your experiments work?

**Comment by varunagrawal** [comment]
This looks great so merging!

**Pull request merged**
